diff --git a/pkgs/applications/editors/emacs/build-support/generic.nix b/pkgs/applications/editors/emacs/build-support/generic.nix
index 3ff10ee86..56d03b6e6 100644
--- a/pkgs/applications/editors/emacs/build-support/generic.nix
+++ b/pkgs/applications/editors/emacs/build-support/generic.nix
@@ -25,7 +25,8 @@ in
 
 libBuildHelper.extendMkDerivation' stdenv.mkDerivation (finalAttrs:
 
-{ buildInputs ? []
+{ version ? null
+, buildInputs ? []
 , nativeBuildInputs ? []
 , packageRequires ? []
 , propagatedBuildInputs ? []
@@ -38,7 +39,7 @@ libBuildHelper.extendMkDerivation' stdenv.mkDerivation (finalAttrs:
 }@args:
 
 {
-  name = args.name or "emacs-${finalAttrs.pname}-${finalAttrs.version}";
+  name = args.name or "emacs-${finalAttrs.pname}${lib.optionalString (version != null) "-${version}"}";
 
   unpackCmd = args.unpackCmd or ''
     case "$curSrc" in
