diff --git a/src/modules/workspaces.rs b/src/modules/workspaces.rs
index 1bafb6d..c448b02 100644
--- a/src/modules/workspaces.rs
+++ b/src/modules/workspaces.rs
@@ -51,20 +51,20 @@ fn get_workspaces(config: &WorkspacesModuleConfig) -> Vec<Workspace> {
     let (special, normal): (Vec<_>, Vec<_>) = workspaces.into_iter().partition(|w| w.id < 0);
 
     // map special workspaces
-    for w in special.iter() {
-        result.push(Workspace {
-            id: w.id,
-            name: w
-                .name
-                .split(":")
-                .last()
-                .map_or_else(|| "".to_string(), |s| s.to_owned()),
-            monitor_id: Some(w.monitor_id as usize),
-            monitor: w.monitor.clone(),
-            active: monitors.iter().any(|m| m.special_workspace.id == w.id),
-            windows: w.windows,
-        });
-    }
+    //for w in special.iter() {
+    //    result.push(Workspace {
+    //        id: w.id,
+    //        name: w
+    //            .name
+    //            .split(":")
+    //            .last()
+    //            .map_or_else(|| "".to_string(), |s| s.to_owned()),
+    //        monitor_id: Some(w.monitor_id as usize),
+    //        monitor: w.monitor.clone(),
+    //        active: monitors.iter().any(|m| m.special_workspace.id == w.id),
+    //        windows: w.windows,
+    //    });
+    //}
 
     // map normal workspaces
     for w in normal.iter() {
