diff --git a/src/modules/workspaces.rs b/src/modules/workspaces.rs
index 806d499..3737a82 100644
--- a/src/modules/workspaces.rs
+++ b/src/modules/workspaces.rs
@@ -41,6 +41,7 @@ fn get_workspaces(enable_workspace_filling: bool) -> Vec<Workspace> {
     let mut workspaces = hyprland::data::Workspaces::get()
         .map(|w| w.to_vec())
         .unwrap_or_default();
+    workspaces.retain(|w| w.id > 0);
 
     workspaces.sort_by_key(|w| w.id);
 
diff --git a/src/outputs.rs b/src/outputs.rs
index adc616b..ce63b57 100644
--- a/src/outputs.rs
+++ b/src/outputs.rs
@@ -58,6 +58,7 @@ impl Outputs {
             id,
             size: Some((None, Some(HEIGHT))),
             layer: Layer::Bottom,
+	    namespace: String::from("ashell"),
             pointer_interactivity: true,
             keyboard_interactivity: KeyboardInteractivity::None,
             exclusive_zone: HEIGHT as i32,
@@ -77,6 +78,7 @@ impl Outputs {
             id: menu_id,
             size: Some((None, None)),
             layer: Layer::Background,
+	    namespace: String::from("ashell"),
             pointer_interactivity: true,
             keyboard_interactivity: KeyboardInteractivity::None,
             output: wl_output.map_or(IcedOutput::Active, |wl_output| {
